Tier1 config cleanup

geooner opened the pull request:
This PR addresses #13483 by implementing the following changes:

1. **Move tier1 configuration from experimental to proper config**: Created a dedicated `Tier1Config` structure in the main `Config` structure, moving tier1 settings out of the experimental section.

2. **Add ability to disable tier1 completely**: Added a new `enabled` field to `Tier1Config` that allows node operators to disable tier1 network completely by setting `tier1.enabled = false`.

The change maintains backward compatibility with existing configurations while providing a cleaner, more maintainable config structure for tier1 network settings.

Fixes #13483

wacban commented:
@saketh-are or @stedfn Can you have a look?

saketh-are commented on chain/network/src/config_json.rs:
```diff
@@ -321,18 +370,19 @@ impl Default for Config {
-            peer_states_cache_size: default_peer_states_cache_size(),
-            snapshot_hosts_cache_size: default_snapshot_hosts_cache_size(),
-            ban_window: Duration::seconds(3 * 60 * 60),
+            ban_window: Duration::hours(1),
```
The previous setting was 3 hours; is the value change intentional?

saketh-are commented on chain/network/src/config_json.rs:
```diff
@@ -231,6 +238,69 @@ fn default_tier1_new_connections_per_attempt() -> u64 {
+/// Configuration for Tier1 network
+///
+/// This replaces the previous configuration that was in the experimental section.
+/// Migration is simple:
```
I think it makes more sense to put the migration instructions in the release notes rather than the source code.

saketh-are commented on chain/network/src/config.rs:
```diff
@@ -288,22 +288,54 @@ impl NetworkConfig {
+        } else {
+            // For backwards compatibility, check if old configuration has tier1 enabled
+            // This is for migration support
+            None
```
The implementation doesn't seem to match the comment above. If we can provide a smooth migration experience that would certainly be great, especially if we can log a warning advising that the experimental config is being deprecated.

However, in practice I don't think we expect or recommend any validator to use a non-standard config for Tier1 network, so I am OK with a more aggressive change.

saketh-are commented on chain/network/src/config_json.rs:
```diff
@@ -231,6 +238,69 @@ fn default_tier1_new_connections_per_attempt() -> u64 {
+pub struct Tier1Config {
+    /// If false, disables Tier1 network completely
+    #[serde(default = "default_tier1_enabled")]
+    pub enabled: bool,
```
Unless I missed something this newly introduced field is not used anywhere.

saketh-are reviewed (changes requested):
Thank you for picking up this task! I left a few comments.

geooner commented:
Hi @saketh-are, I've updated the PR to address your feedback:

*   **`ban_window` duration:** I probably missed the 3 hour, and only saw 60 * 60 for one hour. I reverted it back to 3 hour as it was before.
*   **Migration Instructions:** Removed the detailed migration steps from the code comments in `config_json.rs` as suggested.
*   **`Tier1Config.enabled` Usage:** The logic in `config.rs` now uses the new `tier1.enabled` flag to determine Tier1 network initialization, and this `enabled` field is indeed being used.

Please let me know if there's anything else!
Thank You!

geooner commented:
Hi @saketh-are @wacban, I've updated the PR to address your feedback:

*   **`ban_window` duration:** I probably missed the 3 hour, and only saw 60 * 60 for one hour. I reverted it back to 3 hour as it was before.
*   **Migration Instructions:** Removed the detailed migration steps from the code comments in `config_json.rs` as suggested.
*   **`Tier1Config.enabled` Usage:** The logic in `config.rs` now uses the new `tier1.enabled` flag to determine Tier1 network initialization, and this `enabled` field is indeed being used.

Please let me know if there's anything else!
Thank You!

wacban commented on chain/network/src/config.rs:
```diff
@@ -288,22 +288,53 @@ impl NetworkConfig {
+            .into_iter()
+            .collect::<blacklist::Blacklist>();
+
+        // Determine Tier1 configuration with migration support
```
@saketh-are Do we need backwards compatibility / migration or can we do without it?

saketh-are replied:
I am OK with skipping backwards-compatible migration as there is no reason anyone would have modified the experimental config.

geooner commented:
@geooner @saketh-are Please let me know if any further changes are needed.

wacban commented:
> @wacban @saketh-are Please let me know if any further changes are needed.

Thanks @geooner! It looks good now. I started CI to see if all the tests are passing - can you do a quick grep for the configs to check if any of the pytests use the old configuration? If so those will need to be migrated to the new one. 

Can you mention this change in the top level CHANGELOG.md, in the unreleased, non-protocol changes section. Something like the following would be nice, maybe polish it a little:
* Moved the Tier1 configuration from experimental to top level config. Note for validators - please update configs accordingly. <#PR + link>

geooner commented:
@saketh-are @wacban I have modified Changelog as asked. Please give it a look.

wacban commented:
@geooner clippy and nextest are still failing in CI, please fix

wacban commented on chain/network/src/config.rs:
```diff
@@ -452,15 +473,7 @@ impl NetworkConfig {
-                enable_inbound: true,
-                enable_outbound: true,
-            }),
+            tier1,
```
@geooner My bad, I was too eager to use default config, can you restore it to the one for tests as it was before?

geooner replied:
@wacban you mean before the cherry pick? https://github.com/near/nearcore/pull/13575/commits/f55fbd2bff57145435ce2a14f7f29846b0c1ae67

wacban replied:
this, and remove the tier1 variable, or set tier1 to this value
```suggestion
            tier1: Tier1 {
                // Interval is very large, so that it doesn't happen spontaneously in tests.
                // It should rather be triggered manually in tests.
                connect_interval: time::Duration::hours(1000),
                new_connections_per_attempt: 10000,
                advertise_proxies_interval: time::Duration::hours(1000),
                enable_inbound: true,
                enable_outbound: true,
            },
```

geooner replied:
🤦for some reason your comment are loading with some delay, I'll fix the commits now.
So, what I'll do is:
1. cherry pick https://github.com/near/nearcore/commit/cd77387594f8518685f41acb07fa8fccb600acfc (again)
2. add the values for the test that were missing
3. fix the format
Sounds good?

wacban reviewed (approved):
LGTM, thank you!

geooner commented:
hi @wacban sorry to ping you again, but is there any reason why this isn't being merged?

wacban commented:
@geooner some other issue sneaked in, probably because of merge:
```
error[E0609]: no field `tier1_connect_interval` on type `ExperimentalConfig`
    --> nearcore/src/config.rs:1196:33
     |
1196 |     config.network.experimental.tier1_connect_interval = Duration::seconds(5);
     |                                 ^^^^^^^^^^^^^^^^^^^^^^ unknown field
     |
     = note: available fields are: `inbound_disabled`, `connect_only_to_boot_nodes`, `skip_sending_tombstones_seconds`, `network_config_overrides`
     
 ```
